### DISCUSSION TITLE
Cutting Down Latency: Can I Keep the Prerecorded API's Accuracy?

### DISCUSSION THREAD

**Discussion by 8ta4** [deepgram/community#518] [Feature Requests]
How can I slash latency big time but still keep the accuracy of the prerecorded API intact?

I'm working on an open-source project called [`say`](https://github.com/8ta4/say), which transcribes voice 24/7.

Here's what I'm aiming for:
- [Accuracy](https://github.com/8ta4/say/blob/28276acd622fd2cd8d0f86568534361927ddf363/DONTREADME.md#accuracy): I want to match the accuracy of the prerecorded API.
- [Latency](https://github.com/8ta4/say/blob/28276acd622fd2cd8d0f86568534361927ddf363/DONTREADME.md#latency): I want to cut down Deepgram's latency to around [300 ms](https://deepgram.com/product/transcription#:~:text=time%20streaming%20lag-,%3C300%20ms,-Not%20available) to keep the overall process under a second..
- [Budget](https://github.com/8ta4/say/blob/28276acd622fd2cd8d0f86568534361927ddf363/DONTREADME.md#budget): Given that this is an open-source project, the solution needs to be cost-effective.

The streaming API doesn't quite hit the mark in terms of accuracy. I was dreaming of transcription that's faster than light, but it turned out to be faster than right.

The main speed bump seems to be the audio file upload with the prerecorded API. The TCP slow start might be the troublemaker. For reference, I'm using [Opus](https://github.com/8ta4/say/blob/28276acd622fd2cd8d0f86568534361927ddf363/DONTREADME.md?plain=1#L157) with [a 24kbps bitrate](https://github.com/8ta4/say/blob/28276acd622fd2cd8d0f86568534361927ddf363/DONTREADME.md?plain=1#L165) and [a gigabit connection in the continental US](https://github.com/8ta4/say/blob/28276acd622fd2cd8d0f86568534361927ddf363/DONTREADME.md?plain=1#L251).

I've experimented with a readable stream to send audio data using Deepgram's Node SDK. The idea was to keep the connection alive as more audio gets recorded. But the connection doesn't stick around.

I've brainstormed a few potential solutions:
- Deepgram's Node SDK and Readable Stream: Maybe Deepgram could tweak their Node SDK to keep the connection alive longer when streaming audio data with the prerecorded API.
- Hybrid API: Deepgram could offer an API that allows for real-time audio streaming to be temporarily stored and then processed using the prerecorded API once the streaming is complete. This could bypass the upload delay.
- On-Premise Hosting: Deepgram has [an enterprise on-premise solution](https://deepgram.com/pricing). But maybe they can offer a cheaper on-premise hosting solution that open-source projects can use. Then I could whip up the hybrid API myself.

These solutions might require significant development on Deepgram's part. So, I'm open to any suggestions.

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by jjmaldonis** [comment]
Hey @8ta4! That's quite the ask! You're referencing the "project management" or "speed cost quality" [triangle](https://www.google.com/search?sca_esv=596583007&q=triangle+speed+cost+quality&tbm=isch&source=lnms&sa=X&ved=2ahUKEwiw4LmjnM6DAxWWDDQIHf1mCpcQ0pQJegQIDBAB&biw=1850&bih=932) - this is a triangle that's useful for identifying priorities, and the conclusion is that prioritizing one or two options results in a sacrifice of the other.

The state of the art technology is moving fast and the world will see significant improvements in all the areas you mentioned over the next 1-5 years.

To address your bullet points, the 1st one is likely the option you're looking for. You mentioned that you want to keep the connection alive longer. Have you taken a look at our Keep Alive functionality? Docs are [here](https://developers.deepgram.com/docs/node-sdk-streaming-transcription#keep-alive) and allow you to keep the connection alive for a long period of time.

**Reply by 8ta4** [reply]
Thanks for the quick and free reply. I guess you can't have it all.

Just to make sure we're on the same page, my focus is on the prerecorded API, not the streaming API.

Here's what I had:

```javascript
const { result, error } = await deepgram.listen.prerecorded.transcribeFile(
  fs.createReadStream("./examples/spacewalk.wav"),
  {
    model: "nova",
  }
);
```
https://github.com/deepgram/deepgram-js-sdk?tab=readme-ov-file#local-files

I tried replacing `fs.createReadStream("./examples/spacewalk.wav")` with my custom readable stream, hoping to keep it open. But the connection seems to close prematurely.

If Deepgram is open to tweaking their Node SDK's prerecorded API, I'd be more than happy to provide detailed feedback and even help reproduce this issue I'm facing. If not, I totally get it.

**Reply by jjmaldonis** [reply]
You're definitely welcome to submit a PR! I would create an Issue in our JS SDK repo first and describe your intention. If you request feedback then our team will weigh in on whether or not they will accept the new feature.
